state-of-k8s-cost-optimization/kube-requests-checker.sh

```diff
@@ -0,0 +1,134 @@
+# limitations under the License.
+
+set -o errexit
+set -o pipefail
+
+# Setup variables
+
+CLI_NAME=${0}
+OPTIONS=$(getopt -o n:l:ih --long namespace:,level:,include-kube-system,help -- "$@")
+EXCLUSIONS="--field-selector metadata.namespace!=kube-system"
+
+eval set -- ${OPTIONS}
+
+# Setup options
+
+while true; do
+  case "$1" in
+    -n|--namespace)
+      NAMESPACE="$2"
+      shift 2
+      ;;
+    -l|--level)
+      LEVEL="$2"
+      shift 2
+      ;;
+    -i|--include-kube-system)
+      INCLUDE_KUBE_SYSTEM=true
+      EXCLUSIONS=""
+      shift 1
+      ;;
+    -h|--help)
+      cat << USAGE
+
+kube-requests-checker - Show containers without requests set for CPU, Memory, or both, in a Kubernetes cluster
+
+Usage: ${CLI_NAME} <options>
+
+-n | --namespace <namespace>   : Check containers in one specific namespace                                     Default: All non "kube-system" namespaces
+-l | --level <level>           : Group results by "pod" or "controller". "pod" also returns Pod QoS.            Default: controller
+-i | --include-kube-system     : Include "kube-system" namespace. Cannot be used when specifying one namespace.                                   
+-h | --help                    : Show CLI usage
+
+Example:
+========
+Show all containers with no requests set in the cluster in all non kube-system namespaces.                      $ ${CLI_NAME}
+Show all containers with no requests set in the cluster, in all namespaces, including kube-system.              $ ${CLI_NAME} -i
+Show all containers with no requests set in namespace foo, and group by pod.                                    $ ${CLI_NAME} -n foo -l pod
+
+
+USAGE
+      exit 0
+      ;;
+    --)
+      shift
+      break
+      ;;
+    *)
+      echo "Invalid option: '$1' - run ${CLI_NAME} -h for help."
+      exit 1
+      ;;    
+  esac
+done
+
+# Validate namespace usage
+
+if [ -v NAMESPACE ] && [ -v INCLUDE_KUBE_SYSTEM ]; then
+   echo "Invalid option: 'namespace' (-n) and 'include-kube-system' (-i) cannot be specified simultaneously." 1>&2
+   exit 1
+fi
+
+# Set namespace
+
+if [ -v NAMESPACE ]; then
+    NAMESPACE="-n=${NAMESPACE}"
+    EXCLUSIONS=""
+else
+    NAMESPACE="--all-namespaces"
+fi
+
+# Set level and get containers from pods/controller objects
+
+if [ "${LEVEL}" = "pod" ]; then
+    json=$(kubectl get pods ${NAMESPACE} ${EXCLUSIONS} \
+            -o json | \
+            jq '[.items[] | {kind: .kind, namespace: .metadata.namespace, name: .metadata.name, podQoS: .status.qosClass, containers: [ .spec.containers[] | {name: .name, cpuRequest: .resources.requests.cpu, cpuLimit: .resources.limits.cpu, memRequest: .resources.requests.memory, memLimit: .resources.limits.memory}]}]')
+else
+    if [ "${LEVEL}" = "" ] || [ "${LEVEL}" = "controller" ]; then
+        LEVEL="controller"
+    else
+       echo "Invalid option: '${LEVEL}' is not valid. Must be 'pod' or 'controller'."
+       exit 1
+    fi
```

Comment:
nit: maybe refactoring it to `if elif else` instead to improve script readability?

Reply:
This is done. Thanks!